allennlp/semparse/contexts/table_question_context.py

```diff
@@ -76,31 +77,116 @@
+        # undefined only when one of the year values is -1, but not both.
+        if not isinstance(other, Date):
+            raise ExecutionError("only compare Dates with Dates")
+        # We're doing an exclusive or below.
+        if (self.year == -1) != (other.year == -1):
+            return False  # comparison undefined
+        # If both years are -1, we proceed.
+        if self.year != other.year:
+            return self.year > other.year
+        # The years are equal and not -1, or both are -1.
+        if self.month == -1 or other.month == -1:
+            return False
+        if self.month != other.month:
+            return self.month > other.month
+        # The months and years are equal and not -1
+        if self.day == -1 or other.day == -1:
+            return False
+        return self.day > other.day
+
+    def __ge__(self, other) -> bool:
+        if not isinstance(other, Date):
+            raise ExecutionError("only compare Dates with Dates")
+        return self > other or self == other
+
+    def __str__(self):
+        if (self.month, self.day) == (-1, -1):
+            # If we have only the year, return just that so that the official evaluator does the
+            # comparison against the target as if both are numbers.
+            return str(self.year)
+        return f"{self.year}-{self.month}-{self.day}"
+
+    def __hash__(self):
+        return hash(str(self))
+
+    @classmethod
+    def make_date(cls, string: str) -> 'Date':
+        year_string, month_string, day_string = string.split("-")
+        year = -1
+        month = -1
+        day = -1
+        try:
+            year = int(year_string)
+        except ValueError:
+            pass
+        try:
+            month = int(month_string)
+        except ValueError:
+            pass
+        try:
+            day = int(day_string)
+        except ValueError:
+            pass
+        return Date(year, month, day)
+
+CellValueType = Union[str, float, Date]  # pylint: disable=invalid-name
+
+
 class TableQuestionContext:
     """
     A barebones implementation similar to
     https://github.com/crazydonkey200/neural-symbolic-machines/blob/master/table/wtq/preprocess.py
     for extracting entities from a question given a table and type its columns with <string> | <date> | <number>
     """
     def __init__(self,
-                 table_data: List[Dict[str, str]],
-                 column_types: Dict[str, str],
+                 table_data: List[Dict[str, CellValueType]],
+                 column_types: Dict[str, Set[str]],
+                 column_names: Set[str],
                  question_tokens: List[Token]) -> None:
         self.table_data = table_data
-        self.column_types = column_types
+        self.column_types: Set[str] = set()
+        self.column_names = column_names
+        for types in column_types.values():
+            self.column_types.update(types)
```

Comment:
This is very confusing.  Why do you have `column_types` and `self.column_types` holding different information?  Maybe this would have been cleared up with better documentation on the parameters of this class, but using different names for these two things would be better, too.